src/bootloader.rs

```diff
@@ -0,0 +1,500 @@
+    let substr_start = substr as *const str as *const u8 as usize;
+    let substr_end = substr_start + substr.len();
+
+    if parent_start <= substr_start && substr_end <= parent_end {
+        Some((substr_start - parent_start)..(substr_end - parent_start))
+    } else {
+        None
+    }
+}
+
+#[test]
+fn test_split_cmdline() {
+    let mut it = split_cmdline(r#"th""ís is my="x"" y" cmdline"#);
+    assert_eq!(it.next(), Some(r#"th""ís"#));
+    assert_eq!(it.next(), Some("is"));
+    assert_eq!(it.next(), Some(r#"my="x"" y""#));
+    assert_eq!(it.next(), Some("cmdline"));
+}
+
+#[derive(Debug)]
+pub struct BootLoaderEntryFile {
+    lines: Vec<String>,
+}
+
+impl BootLoaderEntryFile {
+    pub fn new(content: &str) -> Self {
+        Self {
+            lines: content.split_terminator('\n').map(String::from).collect(),
+        }
+    }
+
+    pub fn get_values<'a>(&'a self, key: &'a str) -> impl Iterator<Item = &'a str> + 'a {
+        self.lines
+            .iter()
+            .filter_map(|line| strip_ble_key(line, key))
+    }
+
+    pub fn get_value(&self, key: &str) -> Option<&str> {
+        self.lines.iter().find_map(|line| strip_ble_key(line, key))
+    }
+
+    /// Adds a kernel command-line argument, possibly replacing a previous value.
+    ///
+    /// arg can be something like "composefs=xyz" but it can also be something like "rw".  In
+    /// either case, if the argument already existed, it will be replaced.
+    pub fn add_cmdline(&mut self, arg: &str) {
+        let key = match arg.find('=') {
+            Some(pos) => &arg[..=pos], // include the '='
+            None => arg,
+        };
+
+        // There are three possible paths in this function:
+        //   1. options line with key= already in it (replace it)
+        //   2. options line with no key= in it (append key=value)
+        //   3. no options line (append the entire thing)
+        for line in &mut self.lines {
+            if let Some(cmdline) = strip_ble_key(line, "options") {
+                let segment = split_cmdline(cmdline).find(|s| s.starts_with(key));
+
+                if let Some(old) = segment {
+                    // 1. Replace existing key
+                    let range = substr_range(line, old).unwrap();
+                    line.replace_range(range, arg);
+                } else {
+                    // 2. Append new argument
+                    line.push(' ');
+                    line.push_str(arg);
+                }
+
+                return;
+            }
+        }
+
+        // 3. Append new "options" line with our argument
+        self.lines.push(format!("options {arg}"));
+    }
+
+    /// Adjusts the kernel command-line arguments by adding a composefs= parameter (if appropriate)
+    /// and adding additional arguments, as requested.
+    pub fn adjust_cmdline(&mut self, composefs: Option<&str>, extra: &[&str]) {
+        if let Some(id) = composefs {
+            self.add_cmdline(&format!("composefs={id}"));
```

Comment:
Not a new thing obviously but per discussion I think we should move the linked digest to a file in the initramfs by default, with a kernel argument as override.

Reply:
Which would be fine for sealed images but complicates the story for people who want to derive an image to add a package or whatever.  They need a mechanism to change the root filesystem without rebuilding the initramfs.